File: python/array_search/parametric_search.py
def can_ship(weights, days, capacity):
    used_days = 1
    current_load = 0

    for weight in weights:
        if current_load + weight > capacity:
            used_days += 1
            current_load = 0
        current_load += weight

    return used_days <= days


def parametric_search(weights, days):
    low = max(weights)
    high = sum(weights)

    while low < high:
        mid = (low + high) // 2

        if can_ship(weights, days, mid):
            high = mid
        else:
            low = mid + 1

    return low

File: java/array_search/ParametricSearch.java
public class ParametricSearch {
    private static boolean canShip(int[] weights, int days, int capacity) {
        int usedDays = 1;
        int currentLoad = 0;

        for (int weight : weights) {
            if (currentLoad + weight > capacity) {
                usedDays++;
                currentLoad = 0;
            }
            currentLoad += weight;
        }

        return usedDays <= days;
    }

    public static int parametricSearch(int[] weights, int days) {
        int low = weights[0];
        int high = 0;

        for (int weight : weights) {
            low = Math.max(low, weight);
            high += weight;
        }

        while (low < high) {
            int mid = low + (high - low) / 2;

            if (canShip(weights, days, mid)) {
                high = mid;
            } else {
                low = mid + 1;
            }
        }

        return low;
    }
}
